### ISSUE TITLE
Session Cookie Missing 'Secure' Attribute

### ISSUE THREAD

**Issue by kayvanaarssen** [pglombardo/PasswordPusher#2887]
We got this in a report (Pentest)
Is the fix: https://docs.pwpush.com/docs/other-configurations/ setting - SECRET_KEY_BASE
Or do we need to do more?

Finding;
>Not including the "secure" attribute in a session cookie introduces a significant security risk. Without this attribute, the cookie is transmitted over both secure (HTTPS) and unsecure (HTTP) connections, making it susceptible to interception by attackers on unsecured networks. This vulnerability exposes sensitive session information, such as user authentication tokens, to potential eavesdropping and unauthorized access. Attackers could exploit this weakness to hijack user sessions, leading to unauthorized account access and potential misuse of personal data. Including the "secure" attribute is crucial to ensure that session cookies are transmitted exclusively over encrypted channels, mitigating the risk of data interception and enhancing overall security.

>A session cookie is a small piece of data that a website temporarily stores on a user's device during their visit. It is commonly used to temporarily store information about the user's session, such as login credentials or preferences. Session cookies are designed to be temporary and are typically deleted when the user closes their browser. The "secure" attribute, when applied to a session cookie, ensures it's only transmitted over secure, encrypted connections (HTTPS). This adds an extra layer of protection against potential security threats during data exchange between the user's browser and the server.

- Enable HTTPS for your website to ensure secure communication.
- Explicitly set the "secure" attribute when creating or updating the session cookie.
- Implement secure practices for storing session cookies on the server side.

**Comment by pglombardo** [comment]
Good point.  Should be fixed in #2902 which I'll test and release soon.

**Comment by kayvanaarssen** [comment]
> Good point. Should be fixed in #2902 which I'll test and release soon.

@pglombardo Great news, sorry I missed your reply 2 days ago #BusyTimes

**Comment by jaychinut** [comment]
Glad to see this being worked on as I also had this concern. I was able to set the secure flag by using FORCE_SSL and assume_ssl (due to the load balancer in front of my setup) in config/environments/production.rb. However, I see that FORCE_SSL is being deprecated soon.

# Assume all access to the app is happening through a SSL-terminating reverse proxy.
# Can be used together with config.force_ssl for Strict-Transport-Security and secure cookies.
config.assume_ssl = ENV.key?("ASSUME_SSL")

# Force all access to the app over SSL, use Strict-Transport-Security, and use secure cookies.
config.force_ssl = ENV.key?("FORCE_SSL")

**Comment by kayvanaarssen** [comment]
@pglombardo Is the setting that @jaychinut suggest enough to get it working?
Of are there more settings we need to set in the `.env`?

**Comment by kayvanaarssen** [comment]
That's a good idea, also if you make it still an option in the environment than people can still choose. And make a best practice with the enabled option for an Secure Environment with NGINX etc.

Keep up the good work! But the secure cookies part - is becoming critical with Pentests etc.

**Comment by EarMaster** [comment]
Any news on this one?

**Comment by pglombardo** [comment]
Hi @EarMaster et. al. - a quick summary on this issue:

I attempted to begin to enable secure cookies in #2805 but it broke quite a few installations that either run internally or behind a proxy so I need to add a configurable option here.  Not done yet.

To mitigate risk in the meantime, make sure to serve Password Pusher over SSL only but even still the following is always true:

1. Cookies are encrypted by Password Pusher
2. Cookies are used for session store data and nothing highly sensitive is put into cookies

The [one known risk](https://github.com/pglombardo/PasswordPusher/security/advisories/GHSA-4fwj-m62q-pp47) is that before you log out, someone can copy the session cookie from your browser cookie store and use it elsewhere within the session timeout limit (default 15 minutes).  But this issue is true even when the secure attribute is set.

So with all that said, we need to set the secure attribute for sure.  I'll try to get back to this soon.

**Comment by Zaisuru** [comment]
Hi @pglombardo ,
I have the same problem on my enterprise, do you have any updates about that ?

**Comment by pglombardo** [comment]
Hi @Zaisuru et. al. - Since when I enabled this globally it broke multiple user environments, I'll see if we can instead add a new option to enable it manually.

**Comment by kayvanaarssen** [comment]
> Hi [@Zaisuru](https://github.com/Zaisuru) et. al. - Since when I enabled this globally it broke multiple user environments, I'll see if I can instead add a new option to enable it manually.


Sure thing depends on how you have setup the Proxy etc. So if this could be an option that would be awesome.

**Comment by pglombardo** [comment]
[v1.58.0](https://github.com/pglombardo/PasswordPusher/releases/tag/v1.58.0) is out with a new option to enable secure cookies.

tldr; `PWP__SECURE_COOKIES=true` or set `secure_cookies: true` in your `settings.yml`.

The option has been added to the configuration documentation:
https://docs.pwpush.com/docs/self-hosted-configuration/

**Comment by kayvanaarssen** [comment]
@pglombardo Awesome! Going to test thanks!

**Comment by kayvanaarssen** [comment]
@pglombardo Are you still building the docker images? Cannot pull the 1.58.0 yet.

**Comment by pglombardo** [comment]
Yeah Docker builds usually take 45 mins.  Sorry I posted right away after tag 😅

**Comment by kayvanaarssen** [comment]
@pglombardo Awesome just tested and works according to DOMSignal; https://domsignal.com/secure-cookie-test

**Comment by pglombardo** [comment]
Excellent and nice tool

**Comment by jaychinut** [comment]
I’m helping test the secure cookie flag in Password Pusher 1.58.0 by running it behind a load balancer (Nginx) that terminates SSL. I’ve tried the following settings:

PWP__SECURE_COOKIES=true
FORCE_SSL=true (toggled on and off)
config.assume_ssl = true in [production.rb]

The site loads initially, but when I attempt to log in or "push" a password, I encounter SSL preflight and fetch errors in the browser console. Are there any additional settings or recommendations for ensuring the secure flag works correctly behind an SSL-terminating proxy?

**Comment by pglombardo** [comment]
Hi @jaychinut - are you sending [these headers](https://docs.pwpush.com/docs/proxies/#required-proxy-headers) to the application from nginx?

**Comment by jaychinut** [comment]
Things are working now.  It turned out to be me using non-standard ports in my docker-compose and that just wasn't handled correctly in my nginx.conf.  I didn't need to send the headers you posted. 

just changing to standard ports in my docker-compose worked for me. 
nginx:
  image: nginx:latest
  ports:
    - "80:80"       # Standard HTTP port
    - "443:443"     # Standard HTTPS port
   
Thanks again.  I've now got a functioning pwpush site with secure cookies running behind SSL terminating proxy
